contracts/staking/TokenStaking.sol

```diff
@@ -0,0 +1,1215 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity 0.8.4;
+
+import "./IStaking.sol";
+import "./StakingProviders.sol";
+import "./IApplication.sol";
+import "../token/T.sol";
+import "../vending/VendingMachine.sol";
+import "../utils/PercentUtils.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/math/Math.sol";
+import "@openzeppelin/contracts/utils/math/SafeCast.sol";
+
+/// @notice Meta staking contract. 3 roles: app, meta staking, app manager
+contract TokenStaking is Ownable, IStaking {
+    using SafeERC20 for T;
+    using PercentUtils for uint256;
+    using SafeCast for uint256;
+
+    enum StakingProvider {
+        NU,
+        KEEP,
+        T
+    }
+
+    struct OperatorInfo {
+        address owner;
+        address payable beneficiary;
+        address authorizer;
+        mapping(address => AppAuthorization) authorizations;
+        address[] authorizedApplications;
+        uint96 nuStake;
+        uint96 keepStake;
+        uint96 tStake;
+        uint256 startTStakingTimestamp;
```

Comment:
It would be possible to pack these into a single storage slot by permitting only one type of legacy stake per operator address:
```
uint96 tStake;
uint96 legacyStake;
StakingProvider legacyStakeType;
uint56 startTStakingTimestamp;
```
This should save around 20k gas when staking T and 15k gas when topping up a T stake with legacy stakes.

If we don't want to deal with the complexity of specifying a legacy stake type, we could still pack `tStake` and `startTStakingTimestamp` into a single storage slot by reducing the latter to e.g. `uint64`, still netting the 20k savings when staking T.